stats/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,282 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package opentelemetry implements opentelemetry instrumentation code for
+// gRPC-Go clients and servers.
+package opentelemetry
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal"
+
+	"go.opentelemetry.io/otel/metric"
+)
+
+var logger = grpclog.Component("otel-plugin")
+
+var canonicalString = internal.CanonicalString.(func(codes.Code) string)
+
+var joinDialOptions = internal.JoinDialOptions.(func(...grpc.DialOption) grpc.DialOption)
+
+// EmptyMetrics represents no metrics. To start from a clean slate if the
+// intended effect is to pick a subset of metrics, use this and add onto it.
+var EmptyMetrics = Metrics{}
+
+// MetricName is a name of a metric.
+type MetricName string
+
+// Metrics is a set of metrics to record. Once created, Metrics is immutable.
+type Metrics struct {
+	// metrics are the set of metrics to initialize.
+	metrics map[MetricName]bool
+}
+
+// Add adds the metrics to the metrics set and returns a new copy with the
+// additional metrics.
+func (m *Metrics) Add(metrics ...MetricName) *Metrics {
+	newMetrics := make(map[MetricName]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		newMetrics[metric] = true
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// Remove removes the metrics from the metrics set and returns a new copy with
+// the metrics removed.
+func (m *Metrics) Remove(metrics ...MetricName) *Metrics {
+	newMetrics := make(map[MetricName]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		delete(newMetrics, metric)
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// MetricsOptions are the metrics options for OpenTelemetry instrumentation.
+type MetricsOptions struct {
+	// MeterProvider is the MeterProvider instance that will be used for access
+	// to Named Meter instances to instrument an application. To enable metrics
+	// collection, set a meter provider. If unset, no metrics will be recorded.
+	// Any implementation knobs (i.e. views, bounds) set in the passed in object
+	// take precedence over the API calls from the interface in this component
+	// (i.e. it will create default views for unset views).
+	MeterProvider metric.MeterProvider
+	// Metrics are the metrics to instrument. Will turn on the corresponding
+	// metric supported by the client and server instrumentation components if
+	// applicable.
+	Metrics Metrics
+	// TargetAttributeFilter is a callback that takes the target string of the
+	// channel and returns a bool representing whether to use target as a label
+	// value or use the string "other". If unset, will use the target string as
+	// is.
+	TargetAttributeFilter func(string) bool
```

Comment:
This is valid on client side only?  Please document it that way, or consider splitting `MetricsOptions` into different ones for client & server.

Reply:
It's just a no-op server side since there's no label for target. Added the line "this only applies for client side metrics". I'd rather not split into two separate MetricsOptions.